### PR TITLE
xapi/tipb: add protobuffer files.

### PR DESCRIPTION
**Pull request by coocood**
All known features are added, can be stable for a while.

### CODE REVIEW

File: xapi/tipb/select.proto [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+package tipb;
+
+option java_multiple_files = true;
+option java_package = "com.pingcap.tidb.tipb";
+option java_outer_classname = "SelectProto";
+
+import "expression.proto";
+import "schema.proto";
+
+// KeyRange is the encoded index key ranges, low is closed, high is open. (low <= x < high)
```
Comment (shenli) [inline xapi/tipb/select.proto]: index key ranges --> index key range

File: xapi/tipb/select.proto [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+// SelectRequest works like a simplified select statement.
+message SelectRequest {
+	// transaction start timestamp.
+	optional int64 start_ts = 1;
+
+	// If table_info is not null, it represents a table scan, index_info would be null.
+	optional TableInfo table_info = 2;
+
+	// If IndexInfo is not null, it represents a index scan, table_info would be null.
```
Comment (shenli) [inline xapi/tipb/select.proto]: If index_info
an index scan

File: xapi/tipb/select.proto [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	optional IndexInfo index_info = 3;
+
+	// fields to be selected, fields type can be column reference for simple scan.
+	// or aggregation function. If no fields specified, only handle will be returned.
```
Comment (shenli) [inline xapi/tipb/select.proto]: How about "select substring(c, 1) from t"? Can we push substring down to coprocessor?

Comment (coocood) [inline xapi/tipb/select.proto]: Yes. the fields can be any expression.

File: xapi/tipb/select.proto [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	// order by clause.
+	repeated ByItem order_by = 11;
+
+	// limit the result to be returned;
+	optional int64 limit = 12;
+}
+
+// values are all in text format.
+message Row {
+	optional bytes handle = 1;
```
Comment (shenli) [inline xapi/tipb/select.proto]: How about row contains multiple rows from multiple tables? Or "select count(*) from t for update" ?

File: xapi/tipb/select.proto [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+message SelectResponse {
+	optional Error error = 1;
+
+	// Result rows.
+	repeated Row rows = 3;
+
+	// If the requested key range has not finished scanning.
+	// next_handle can be used to compose the range for the following request.
+	// null means there is no more data to scan.
+	optional bytes next_handle = 4;
```
Comment (shenli) [inline xapi/tipb/select.proto]: XAPI does not know the range distribution.

File: xapi/tipb/select.proto (hunk 4 above)
Comment (coocood) [inline xapi/tipb/select.proto]: We can restrict `for update` usage, only support it for non-aggregate query.

**Comment by shenli** [comment]
@nieyy PTAL

**Comment by nieyy** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

File: xapi/tipb/schema.proto [hunk 6]
```diff
@@ -0,0 +1,30 @@
+syntax = "proto2";
+
+package tipb;
+
+option java_multiple_files = true;
+option java_package = "com.pingcap.tidb.tipb";
+option java_outer_classname = "SelectProto";
```
Comment (zimulala) [inline xapi/tipb/schema.proto]: SelectProto -> SchemaProto ?

Comment (coocood) [inline xapi/tipb/schema.proto]: According to https://developers.google.com/protocol-buffers/docs/javatutorial#why-use-protocol-buffers
This option can be absent, then the class name would be camel case file name.
I'll remove it.

File: xapi/tipb/select.proto [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	optional int32 code = 1;
+	optional string msg = 2;
+}
+
+// Response for SelectRequest.
+message SelectResponse {
+	optional Error error = 1;
+
+	// Result rows.
+	repeated Row rows = 3;
```
Comment (zimulala) [inline xapi/tipb/select.proto]: 3 -> 2 ?

**Comment by zimulala** [comment]
LGTM

**Pull request merged**